### PR TITLE
[move-prover] multiple fixes to the memory model

### PR DESCRIPTION
**Pull request by meng-xu-cs**
This PR contains a chain of commits that fixes several issues in the memory model:

The best way to review this PR is to review the commits one by one in the following order:
[[move-prover] fix is_parent tests for multiple write-back chains](https://github.com/move-language/move/pull/312/commits/a8c535c4eeba85216dc748a7ff71d7c109db92f9)
[[move-prover] improve the peephole optimization for write-back chains](https://github.com/move-language/move/pull/312/commits/0f651a6753a73c8f92ec5fad082b16952cd50166)
[[move-prover] do not peephole optimize the initial local and global borrows](https://github.com/move-language/move/commit/45b24d5571ca425c5684ee45b08e6849bb8dae81)
[[move-prover] use a special location to mark uninitialized mut refs](https://github.com/move-language/move/commit/075d51565ee91db1eb43652a4815f2476f2f37a4)
[[move-prover] use an Uninit operation to mark uninitialized mut refs](https://github.com/move-language/move/commit/a0acb5d662b2751069ab8c15a72f660ea44bb9d9)
[[move-prover] do not let bind a mutable reference for EmitEvent](https://github.com/move-language/move/commit/1b88dbeb1d4e576496eaa4636071b72cd9a78b4c)

The last two commits are updates to test cases.

The commit message contains much more details, including what is the issue and a
high level summary of the fix.

## Motivation

Improve the memory modeling.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/move-language/move/blob/main/CONTRIBUTING.md#developer-workflow)?

Yes

## Test Plan

- CI
- New test cases added
- DPN inconsistency test

### CODE REVIEW

File: language/move-prover/bytecode/tests/memory_instr/basic_test.exp [hunk 1]
```diff
@@ -401,64 +409,130 @@ fun TestPackref::test8($t0|b: bool, $t1|n: u64, $t2|r_ref: &mut TestPackref::R)
...
+ 42: label L5
+ 43: $t16 := borrow_local($t4)
+ 44: $t5 := $t16
+ 45: label L6
+ 46: $t17 := 1
+ 47: $t1 := -($t1, $t17)
+ 48: goto 6
+ 49: label L2
+ 50: if ($t0) goto 51 else goto 77
+ 51: label L8
```
Comment (wrwg) [inline language/move-prover/bytecode/tests/memory_instr/basic_test.exp:484]: Wow thats a lot more code

Comment (meng-xu-cs) [reply language/move-prover/bytecode/tests/memory_instr/basic_test.exp:484]: Yes, it looks rather suspicious... The duplication is even more obvious in the [CFG](https://github.com/move-language/move/files/9214035/g.pdf).

I have a guess on the reason, but need some more time to verify it and fix it.

File: language/move-prover/bytecode/src/livevar_analysis.rs [hunk 2]
```diff
@@ -252,7 +276,9 @@ impl<'a> LiveVarAnalysis<'a> {
                     // Drop this load/assign as it is not used.
                 }
                 Bytecode::Call(attr_id, dests, oper, srcs, aa)
-                    if code_offset + 1 < code.len() && dests.len() == 1 =>
+                    if code_offset + 1 < code.len()
+                        && dests.len() == 1
+                        && !matches!(oper, Operation::BorrowLoc | Operation::BorrowGlobal(..)) =>
```
Comment (meng-xu-cs) [inline language/move-prover/bytecode/src/livevar_analysis.rs:281]: Unfortunately, this is the culprit for the lengthy code in the `exp` files in `move-prover/bytecode/tests`.

And either
1) we keep this (i.e., drop the optimization on immediately dead borrows), such that we maintain the invariant that the `x` and `y` in `IsParent[x](y)` are both references, or
2) we support `IsParent[LocalRoot(x)](y)` and `IsParent[GlobalRoot(x)](y)`...

To me 1) is preferred due to simplicity in code and in concept, but let me know if we want to go to route 2).

File: language/move-prover/bytecode/tests/memory_instr/basic_test.exp (hunk 1 above)
Comment (meng-xu-cs) [reply language/move-prover/bytecode/tests/memory_instr/basic_test.exp:484]: The reason for the code explosion is explained below.

And I fixed the `clean_and_optimize` pass such that most of the code is optimized away in the next phase, as seen in the 
[CFG](https://github.com/move-language/move/files/9222771/a-8.pdf) here

**Review by wrwg — APPROVED** [review]

**Pull request merged**